lib/internal/streams/end-of-stream.js

```diff
@@ -28,6 +28,19 @@ function eos(stream, opts, callback) {
 
   callback = once(callback);
 
+  const onerror = (err) => {
+    callback.call(stream, err);
+  };
+
+  if (stream.destroyed) {
```

Comment:
@mcollina: spending some more time thinking about this... I think we should also check for `readableEnded` and `writableFinished`. Not all streams will emit e.g. `'close'` after `'end'` or `'finish'` which means that this PR only partly fixes the problem and some "incorrect" streams can still deadlock.

Reply:
We slightly encountered this here https://github.com/nodejs/node/pull/29176

Reply:
I'm definitely +1 in checking `readableEnded` and `writableFinished` here.

Reply:
@mcollina: fixed, please take another look.